File: web/src/lib/dsl.ts
// `ctx` API for script actions (see docs/dsl.md, copied into help/dsl.md):
// powers the code editor's autocompletion.

export type DslGroup = 'read' | 'write' | 'call';

export interface DslFunction {
  /** JavaScript name (camelCase); the Go name is PascalCase */
  name: string;
  args: string[];
  returns: string;
  doc: string;
  group: DslGroup;
}

export const DSL_FUNCTIONS: DslFunction[] = [
  { name: 'intent', args: [], returns: 'string', doc: "Intent expressed at the origin of the process.", group: 'read' },
  { name: 'goal', args: [], returns: 'string', doc: 'Goal being pursued.', group: 'read' },
  { name: 'agent', args: [], returns: 'string', doc: "Agent executing the action.", group: 'read' },
  { name: 'action', args: [], returns: 'string', doc: "Name of the current action.", group: 'read' },
  { name: 'param', args: ['name'], returns: 'JSON value', doc: "Action parameter.", group: 'read' },
  { name: 'var', args: ['name'], returns: 'JSON value', doc: 'Process variable (vars from StartProcess).', group: 'read' },
  { name: 'items', args: ['kind'], returns: 'Item[]', doc: 'Items of the change of this kind ("" = all).', group: 'read' },
  { name: 'impacts', args: [], returns: 'Item[]', doc: 'Impacts of the change.', group: 'read' },
  { name: 'proposals', args: [], returns: 'Item[]', doc: 'Proposals of the change.', group: 'read' },
  { name: 'node', args: ['key'], returns: 'Node', doc: 'Node from the reference baseline.', group: 'read' },
  { name: 'nodes', args: ['type'], returns: 'Node[]', doc: 'Nodes of this type ("" = all).', group: 'read' },
  {
    name: 'links',
    args: ['key', 'direction', 'type'],
    returns: 'Link[]',
    doc: 'Links of a node; direction "out" / "in", type "" = all.',
    group: 'read',
  },
  { name: 'addImpact', args: ['key', 'reason'], returns: '', doc: 'Direct impact on a baseline node.', group: 'write' },
  {
    name: 'proposeNode',
    args: ['type', 'key', 'props'],
    returns: '"#pN"',
    doc: 'Proposal to create a node; returns a #pN reference.',
    group: 'write',
  },
  { name: 'proposeUpdate', args: ['key', 'props'], returns: '', doc: "New version of a node.", group: 'write' },
  { name: 'proposeDelete', args: ['key'], returns: '', doc: "Deletion of a node.", group: 'write' },
  {
    name: 'proposeLink',
    args: ['from', 'type', 'to'],
    returns: '',
    doc: 'Link between two nodes (node key or #pN reference).',
    group: 'write',
  },
  { name: 'addArtifact', args: ['type', 'data'], returns: '', doc: 'Free-form data (report…).', group: 'write' },
  { name: 'decide', args: ['itemId', 'accept', 'comment'], returns: '', doc: 'Decision on a proposal.', group: 'write' },
  { name: 'llm', args: ['prompt'], returns: 'string', doc: 'Text completion ("default" model).', group: 'call' },
  {
    name: 'complete',
    args: ['{model, system, prompt, json, maxTokens}'],
    returns: '{text, json, inputTokens, outputTokens, model}',
    doc: 'Completion with options.',
    group: 'call',
  },
  {
    name: 'runAgent',
    args: ['name', 'intent'],
    returns: '{status, goal, processId}',
    doc: "Runs a sub-agent on the same change; if the sub-agent waits on a human, the action is suspended then replayed.",
    group: 'call',
  },
  { name: 'callTool', args: ['name', 'args'], returns: 'result', doc: 'MCP tool (server/tool).', group: 'call' },
  { name: 'log', args: ['msg'], returns: '', doc: "Log entry (IDE console).", group: 'call' },
  { name: 'warn', args: ['msg'], returns: '', doc: "Warning (IDE console).", group: 'call' },
];

/** Fields of the returned objects (for completion after `i.` or `n.`). */
export const DSL_FIELDS: Record<string, string[]> = {
  Item: ['id', 'kind', 'type', 'status', 'target', 'data', 'op', 'node', 'producedBy'],
  Node: ['id', 'version', 'key', 'type', 'state', 'props'],
  Link: ['id', 'type', 'from', 'to'],
};

export function goName(name: string): string {
  return name.charAt(0).toUpperCase() + name.slice(1);
}

export const TEMPLATES: Record<string, string> = {
  javascript: `// Script actions: the ctx object gives access to the blackboard (see "DSL Help").
for (const i of ctx.impacts()) {
  ctx.log("impact on " + i.target.key);
}
`,
  go: `package action

import "github.com/zimwip/goap/pkg/dsl"

func Run(ctx *dsl.Ctx) error {
\tfor _, i := range ctx.Impacts() {
\t\tctx.Log("impact on " + i.Target.Key)
\t}
\treturn nil
}
`,
};

// --- algorithms (ADR 0018) -----------------------------------------------------------
// Algorithms plug into the model at fixed extension points. Each type runs in its own
// `ctx`: a JavaScript algorithm is the BODY of a function of ctx (it may `return`), a Go
// algorithm declares `func Run(ctx *dsl.<Ctx>) error`. They are pure: no blackboard, LLM or tool calls.

export type AlgorithmUsage = 'property_validator' | 'transition_guard' | 'transition_action';

export interface AlgorithmUsageInfo {
  usage: AlgorithmUsage;
  title: string;
  /** Go context type */
  goCtx: string;
  description: string;
  /** what the plug does with the result */
  contract: string;
  functions: DslFunction[];
  /** sample input of "Try it" */
  sample: Record<string, unknown>;
}

const paramFn: DslFunction = { name: 'param', args: ['name'], returns: 'JSON value', doc: 'Parameter value set by the instance (defaults filled in).', group: 'read' };
const failFn: DslFunction = { name: 'fail', args: ['message'], returns: '', doc: 'Reject with a message (may be called several times).', group: 'call' };
const logFns: DslFunction[] = [
  { name: 'log', args: ['msg'], returns: '', doc: 'Log entry.', group: 'call' },
  { name: 'warn', args: ['msg'], returns: '', doc: 'Warning.', group: 'call' },
];
const metaFns: DslFunction[] = [
  { name: 'instance', args: [], returns: 'string', doc: 'Name of the algorithm instance being run.', group: 'read' },
  { name: 'algorithm', args: [], returns: 'string', doc: 'Name of the algorithm.', group: 'read' },
];
const nodeFn: DslFunction = { name: 'node', args: [], returns: 'Node', doc: 'The node concerned: {id, version, key, type, state, props}.', group: 'read' };
const groupFns: DslFunction[] = [
  { name: 'children', args: [], returns: 'Node[]', doc: 'Nodes a document contains, as they are in the result of the change.', group: 'read' },
  { name: 'change', args: [], returns: '{id, title, intent, methodology, goal}', doc: 'The change being applied.', group: 'read' },
  { name: 'transition', args: [], returns: '{name, from, to}', doc: 'The lifecycle transition.', group: 'read' },
];

export const ALGORITHM_USAGES: AlgorithmUsageInfo[] = [
  {
    usage: 'property_validator',
    title: 'Property validator',
    goCtx: 'ValidatorCtx',
    description: 'Checks the value of a property when a node of the type is created or modified.',
    contract: 'Rejects with ctx.fail(msg), a throw (Go: a returned error), or by returning false / a message (JavaScript).',
    functions: [
      paramFn,
      { name: 'property', args: [], returns: 'string', doc: 'Name of the property being validated.', group: 'read' },
      { name: 'value', args: [], returns: 'JSON value', doc: 'Value of the property (null when the node has none).', group: 'read' },
      nodeFn,
      failFn,
      ...metaFns,
      ...logFns,
    ],
    sample: { property: 'title', value: 'Card payment', node: { key: 'REQ-1', type: 'Requirement', state: 'draft', props: { title: 'Card payment' } } },
  },
  {
    usage: 'transition_guard',
    title: 'Transition guard',
    goCtx: 'GuardCtx',
    description: 'Decides whether a lifecycle transition is allowed; checked when the change is applied, after the CEL guard.',
    contract: 'Refuses with ctx.fail(msg), a throw / returned error, or by returning false / a message (JavaScript).',
    functions: [paramFn, nodeFn, ...groupFns, failFn, ...metaFns, ...logFns],
    sample: {
      node: { key: 'SPEC-1', type: 'Spec', state: 'released', props: { title: 'Spec' } },
      children: [{ key: 'REQ-1', type: 'Requirement', state: 'approved', props: {} }, { key: 'REQ-2', type: 'Requirement', state: 'draft', props: {} }],
      transition: { name: 'release', from: 'draft', to: 'released' },
      change: { id: 'c1', title: 'Release the spec' },
    },
  },
  {
    usage: 'transition_action',
    title: 'Transition action',
    goCtx: 'TransitionCtx',
    description: 'Runs once a transition is accepted and may change properties of the node in the version it produces.',
    contract: 'Changes properties with ctx.setProp / ctx.removeProp; ctx.fail(msg) or a throw aborts the application of the change.',
    functions: [
      paramFn,
      nodeFn,
      ...groupFns,
      { name: 'setProp', args: ['name', 'value'], returns: '', doc: 'Set a property of the node.', group: 'write' },
      { name: 'removeProp', args: ['name'], returns: '', doc: 'Remove a property of the node.', group: 'write' },
      failFn,
      ...metaFns,
      ...logFns,
    ],
    sample: {
      node: { key: 'REQ-1', type: 'Requirement', state: 'approved', props: { title: 'Card payment' } },
      transition: { name: 'approve', from: 'in_review', to: 'approved' },
      change: { id: 'c1', title: 'Approve REQ-1' },
    },
  },
];

export function algorithmUsage(usage: string): AlgorithmUsageInfo | undefined {
  return ALGORITHM_USAGES.find((u) => u.usage === usage);
}

export const ALGORITHM_TEMPLATES: Record<AlgorithmUsage, Record<'javascript' | 'go', string>> = {
  property_validator: {
    javascript: `// Body of a function of ctx. Reject with ctx.fail(message) or by returning a message.
const v = ctx.value();
if (typeof v !== "string" || v.trim() === "") {
  ctx.fail(ctx.property() + " must not be blank");
}
`,
    go: `package validator

import (
\t"strings"

\t"github.com/zimwip/goap/pkg/dsl"
)

func Run(ctx *dsl.ValidatorCtx) error {
\ts, _ := ctx.Value().(string)
\tif strings.TrimSpace(s) == "" {
\t\tctx.Fail(ctx.Property() + " must not be blank")
\t}
\treturn nil
}
`,
  },
  transition_guard: {
    javascript: `// Body of a function of ctx. Refuse with ctx.fail(message).
for (const c of ctx.children()) {
  if (c.state !== "approved") ctx.fail(c.key + " is " + c.state);
}
`,
    go: `package guard

import "github.com/zimwip/goap/pkg/dsl"

func Run(ctx *dsl.GuardCtx) error {
\tfor _, c := range ctx.Children() {
\t\tif c.State != "approved" {
\t\t\tctx.Fail(c.Key + " is " + c.State)
\t\t}
\t}
\treturn nil
}
`,
  },
  transition_action: {
    javascript: `// Body of a function of ctx. Change the node with ctx.setProp / ctx.removeProp.
ctx.setProp("lastTransition", ctx.transition().name);
`,
    go: `package action

import "github.com/zimwip/goap/pkg/dsl"

func Run(ctx *dsl.TransitionCtx) error {
\tctx.SetProp("lastTransition", ctx.Transition().Name)
\treturn nil
}
`,
  },
};

File: pkg/dsl/dsl.go
// Package dsl is the API injected in script actions (JavaScript or Go): it
// reads the blackboard snapshot and the reference domain, buffers writes to
// the change, and calls the platform (LLM, sub-agents, tools) through a Host.
// The same Go type backs both languages: methods are exposed in camelCase to
// JavaScript (goja) and as is to Go (yaegi). See docs/dsl.md.
package dsl

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/zimwip/goap/pkg/domain"
)

// Node is a domain node version.
type Node struct {
	ID      string `json:"id"`
	Version int    `json:"version"`
	Key     string `json:"key"`
	Type    string `json:"type"`
	// State in the lifecycle of the node type (empty: none).
	State string         `json:"state"`
	Props map[string]any `json:"props"`
}

// LinkEnd is a link endpoint summary.
type LinkEnd struct {
	ID      string `json:"id"`
	Version int    `json:"version"`
	Key     string `json:"key"`
	Type    string `json:"type"`
}

// Link is a domain link.
type Link struct {
	ID   string  `json:"id"`
	Type string  `json:"type"`
	From LinkEnd `json:"from"`
	To   LinkEnd `json:"to"`
}

// Item is a change item of the blackboard snapshot.
type Item struct {
	ID         string `json:"id"`
	Kind       string `json:"kind"`
	Type       string `json:"type"`
	Status     string `json:"status"`
	Op         string `json:"op"`
	ProducedBy string `json:"producedBy"`
	Target     *Node  `json:"target"`
	Node       *Node  `json:"node"` // proposal node (draft), base node when updating
	// Link is the link of add_link / remove_link proposals.
	Link *ItemLink      `json:"link"`
	Data map[string]any `json:"data"`
}

// ItemLink is a proposed link: endpoints are node keys, or "@<itemId>" for
// nodes proposed by other items (usable as is in ProposeLink).
type ItemLink struct {
	ID   string `json:"id"`
	Type string `json:"type"`
	From string `json:"from"`
	To   string `json:"to"`
}

// CompleteRequest is an LLM request.
type CompleteRequest struct {
	Model     string `json:"model"`
	System    string `json:"system"`
	Prompt    string `json:"prompt"`
	JSON      bool   `json:"json"`
	MaxTokens int    `json:"maxTokens"`
}

// CompleteResult is an LLM answer.
type CompleteResult struct {
	Text         string `json:"text"`
	JSON         any    `json:"json"`
	Model        string `json:"model"`
	InputTokens  int    `json:"inputTokens"`
	OutputTokens int    `json:"outputTokens"`
}

// AgentResult is the outcome of a sub-agent run.
type AgentResult struct {
	Status    string `json:"status"`
	Goal      string `json:"goal"`
	ProcessID string `json:"processId"`
}

// LogLine is a script log line.
type LogLine struct {
	Time    time.Time `json:"time"`
	Level   string    `json:"level"`
	Message string    `json:"message"`
}

// ErrSuspended is returned by Host.RunAgent when the sub-agent is blocked
// (waiting for a human): the action is suspended and retried later.
var ErrSuspended = errors.New("suspended: waiting for a sub-agent")

// Host performs the operations that leave the script: in a sandbox it is a
// client of the engine RuntimeService.
type Host interface {
	Complete(ctx context.Context, req CompleteRequest) (CompleteResult, error)
	RunAgent(ctx context.Context, name, intent string) (AgentResult, error)
	CallTool(ctx context.Context, name string, args map[string]any) (any, error)
	Node(ctx context.Context, key string) (Node, error)
	Nodes(ctx context.Context, nodeType string) ([]Node, error)
	Links(ctx context.Context, key, direction, linkType string) ([]Link, error)
}

// Job is a script execution.
type Job struct {
	Language  string         `json:"language"`
	Code      string         `json:"code"`
	ProcessID string         `json:"processId"`
	Agent     string         `json:"agent"`
	Action    string         `json:"action"`
	Intent    string         `json:"intent"`
	Goal      string         `json:"goal"`
	Params    map[string]any `json:"params"`
	Vars      map[string]any `json:"vars"`
	Items     []Item         `json:"items"`
	Timeout   time.Duration  `json:"timeout"`
}

// Result is the outcome of a script: the items to add to the change (engine
// ItemInput format), logs and whether the script was suspended.
type Result struct {
	Items     []map[string]any `json:"items"`
	Output    string           `json:"output"`
	Logs      []LogLine        `json:"logs"`
	Suspended bool             `json:"suspended"`
}

// Ctx is the object injected as `ctx` in scripts.
type Ctx struct {
	job       Job
	host      Host
	gctx      context.Context
	out       []map[string]any
	logs      []LogLine
	seq       int
	suspended bool
	result    any
}

func newCtx(gctx context.Context, job Job, host Host) *Ctx {
	return &Ctx{job: job, host: host, gctx: gctx}
}

// ---- context ----------------------------------------------------------------

func (c *Ctx) Intent() string { return c.job.Intent }
func (c *Ctx) Goal() string   { return c.job.Goal }
func (c *Ctx) Agent() string  { return c.job.Agent }
func (c *Ctx) Action() string { return c.job.Action }

// Param returns an action parameter.
func (c *Ctx) Param(name string) any { return c.job.Params[name] }

// Var returns a process variable.
func (c *Ctx) Var(name string) any { return c.job.Vars[name] }

// ---- blackboard (read) ----------------------------------------------------

// Items returns the change items of a kind ("" = all).
func (c *Ctx) Items(kind string) []Item {
	out := []Item{}
	for _, it := range c.job.Items {
		if kind == "" || it.Kind == kind {
			out = append(out, it)
		}
	}
	return out
}

func (c *Ctx) Impacts() []Item   { return c.Items(string(domain.KindImpact)) }
func (c *Ctx) Proposals() []Item { return c.Items(string(domain.KindProposal)) }

// ---- domain (read, reference baseline) --------------------------------------

func (c *Ctx) Node(key string) (Node, error) { return c.host.Node(c.gctx, key) }
func (c *Ctx) Nodes(nodeType string) ([]Node, error) {
	return c.host.Nodes(c.gctx, nodeType)
}
func (c *Ctx) Links(key, direction, linkType string) ([]Link, error) {
	return c.host.Links(c.gctx, key, direction, linkType)
}

// ---- change (buffered writes) ---------------------------------------------

func (c *Ctx) emit(item map[string]any) string {
	c.seq++
	ref := fmt.Sprintf("p%d", c.seq)
	item["ref"] = ref
	item["producedBy"] = c.job.Action
	c.out = append(c.out, item)
	return "#" + ref
}

// AddImpact records a direct impact on a node of the reference baseline.
func (c *Ctx) AddImpact(key, reason string) string {
	return c.emit(map[string]any{"kind": "impact", "type": "direct", "target": key, "data": map[string]any{"reason": reason}})
}

// ProposeNode proposes a new node and returns its reference ("#pN").
func (c *Ctx) ProposeNode(nodeType, key string, props map[string]any) string {
	return c.emit(map[string]any{"kind": "proposal", "proposal": map[string]any{"op": "create_node",
		"node": map[string]any{"type": nodeType, "key": key, "props": props}}})
}

// ProposeUpdate proposes a new version of a node.
func (c *Ctx) ProposeUpdate(key string, props map[string]any) string {
	return c.emit(map[string]any{"kind": "proposal", "proposal": map[string]any{"op": "update_node",
		"node": map[string]any{"base": key, "props": props}}})
}

// ProposeDelete proposes the deletion of a node.
func (c *Ctx) ProposeDelete(key string) string {
	return c.emit(map[string]any{"kind": "proposal", "proposal": map[string]any{"op": "delete_node", "node": map[string]any{"base": key}}})
}

// ProposeTransition proposes to move a node to a lifecycle state. A node of a
// type with a lifecycle is only modified in an editable state: reopen it with a
// transition first, and move it out of the editable states at the end.
func (c *Ctx) ProposeTransition(key, state string) string {
	return c.emit(map[string]any{"kind": "proposal", "proposal": map[string]any{"op": "transition_node",
		"node": map[string]any{"base": key, "state": state}}})
}

// ProposeLink proposes a link; from and to are node keys or "#pN" references.
func (c *Ctx) ProposeLink(from, linkType, to string) string {
	return c.emit(map[string]any{"kind": "proposal", "proposal": map[string]any{"op": "add_link",
		"link": map[string]any{"type": linkType, "from": from, "to": to}}})
}

// AddArtifact records free data (report…).
func (c *Ctx) AddArtifact(artifactType string, data map[string]any) string {
	return c.emit(map[string]any{"kind": "artifact", "type": artifactType, "data": data})
}

// Decide accepts or rejects a proposal (item id, or "#pN").
func (c *Ctx) Decide(item string, accept bool, comment string) string {
	if len(item) > 0 && item[0] != '#' && item[0] != '@' {
		item = "@" + item
	}
	return c.emit(map[string]any{"kind": "decision", "decision": map[string]any{"item": item, "accept": accept, "comment": comment}})
}

// ---- platform calls ---------------------------------------------------------

// LLM completes a prompt with the default model and returns the text.
func (c *Ctx) LLM(prompt string) (string, error) {
	r, err := c.Complete(CompleteRequest{Prompt: prompt})
	return r.Text, err
}

// Complete calls the model gateway.
func (c *Ctx) Complete(req CompleteRequest) (CompleteResult, error) {
	if req.Model == "" {
		req.Model = "default"
	}
	r, err := c.host.Complete(c.gctx, req)
	if err != nil {
		return r, err
	}
	if req.JSON && r.JSON == nil {
		var v any
		if json.Unmarshal([]byte(extractJSON(r.Text)), &v) == nil {
			r.JSON = v
		}
	}
	return r, nil
}

// RunAgent runs a sub-agent on the same change.
func (c *Ctx) RunAgent(name, intent string) (AgentResult, error) {
	r, err := c.host.RunAgent(c.gctx, name, intent)
	if errors.Is(err, ErrSuspended) {
		c.suspended = true
	}
	return r, err
}

// CallTool calls an MCP tool ("server/tool").
func (c *Ctx) CallTool(name string, args map[string]any) (any, error) {
	return c.host.CallTool(c.gctx, name, args)
}

// ---- logs -------------------------------------------------------------------

func (c *Ctx) logf(level, msg string) {
	c.logs = append(c.logs, LogLine{Time: time.Now().UTC(), Level: level, Message: msg})
}

func (c *Ctx) Log(msg string)  { c.logf("info", msg) }
func (c *Ctx) Warn(msg string) { c.logf("warn", msg) }

// SetOutput sets the human readable output of the action.
func (c *Ctx) SetOutput(v any) { c.result = v }

func extractJSON(s string) string {
	for i, r := range s {
		if r == '{' || r == '[' {
			return s[i:]
		}
	}
	return s
}

// ItemsFromBlackboard builds the item snapshot given to scripts.
func ItemsFromBlackboard(bb domain.Blackboard) []Item {
	view := func(r *domain.NodeRef) *Node {
		if r == nil {
			return nil
		}
		n := &Node{ID: string(r.ID), Version: int(r.Version)}
		if v, ok := bb.Nodes[*r]; ok {
			n.Key, n.Type, n.Props = v.Key, v.Type, v.Properties
		}
		return n
	}
	out := make([]Item, 0, len(bb.Change.Items))
	for _, it := range bb.Change.Items {
		if !bb.Change.Active(it.ID) {
			continue
		}
		x := Item{ID: string(it.ID), Kind: string(it.Kind), Type: it.Type, Status: string(bb.Change.EffectiveStatus(it.ID)),
			ProducedBy: it.ProducedBy, Target: view(it.Target), Data: it.Data}
		if p := it.Proposal; p != nil {
			x.Op = string(p.Op)
			if l := p.Link; l != nil {
				end := func(e domain.Endpoint) string {
					if e.Item != "" {
						return "@" + string(e.Item)
					}
					if n := view(e.Node); n != nil && n.Key != "" {
						return n.Key
					}
					return e.String()
				}
				x.Link = &ItemLink{ID: string(l.LinkID), Type: l.Type, From: end(l.From), To: end(l.To)}
			}
			if p.Node != nil {
				x.Node = &Node{Key: p.Node.Key, Type: p.Node.Type, Props: p.Node.Properties}
				if b := view(p.Node.Base); b != nil {
					x.Node.ID, x.Node.Version = b.ID, b.Version
					if x.Node.Key == "" {
						x.Node.Key = b.Key
					}
					if x.Node.Type == "" {
						x.Node.Type = b.Type
					}
				}
			}
		}
		out = append(out, x)
	}
	return out
}
